protos/isspbftpb/isspbftpb.proto

```diff
@@ -36,34 +36,39 @@ message Commit {
 }
 
 message ViewChange {
-  uint64 view = 1;
-  map<uint64, Prepare> pset = 2;
-  map<uint64, Prepare> qset = 3;
+  uint64             view       = 1;
+  repeated PSetEntry p_set      = 2;
+  repeated QSetEntry q_set      = 3;
 }
 
-// ============================================================
-// Events
-// ============================================================
-
-message ReqWaitReference {
-  uint64   sn = 1;
-  uint64 view = 2;
+message SignedViewChange {
+  ViewChange view_change = 1;
+  bytes      signature   = 2;
 }
 
-message PersistPreprepare {
-  Preprepare preprepare = 1;
+message PSetEntry {
+  uint64 sn     = 1;
+  uint64 view   = 2;
+  bytes  digest = 3;
 }
 
-message PersistPrepare {
-  Prepare prepare = 1;
+message QSetEntry {
+  uint64 sn     = 1;
+  uint64 view   = 2;
+  bytes  digest = 3;
 }
 
-message PersistCommit {
-  Commit commit = 1;
+message PreprepareRequest {
+  bytes digest = 1;
 }
 
-message PreprepareHashOrigin {
-  Preprepare preprepare = 1;
+// ============================================================
+// Events
+// ============================================================
+
+message ReqWaitReference {
+  uint64   sn = 1;
+  uint64 view = 2;
```

Comment:
What is this for?

Reply:
It's part of the PBFT code that uses signatures for view changes. It's meant to make sure that requests are properly preprepared across views. It is not directly linked to signatures themselves, but the PBFT algorithm (after introducing signatures and view changes) needs it.